BUG: endpoints of array returned by geomspace() should match arguments

kasparthommen opened the pull request:
The function `numpy.geomspace()` does not respect boundaries exactly, see example below:

```python
import numpy as np

start = 0.3
end = 20.3
gs = np.geomspace(start, end, num=3)
print(start, gs[0])  # prints: 0.3 0.29999999999999993
print(end, gs[-1])  # prints: 20.3 20.300000000000004
```

I think `geomspace()` should explicitly correct the first and the last entry of the generated output array because `np.exp(np.log(x))` is not necessarily equal to `x`. This pull request represents a fix including unit tests.

kasparthommen committed:
BUG: make sure the endpoints of the array returned by geomspace() matches the 'start' and 'stop' arguments exactly

eric-wieser commented on numpy/core/function_base.py:
```diff
@@ -408,8 +408,18 @@ def geomspace(start, stop, num=50, endpoint=True, dtype=None, axis=0):
+    if num > 0:
+        result[0] = start
+        if num > 1 and endpoint:
+            result[-1] = stop
```
This code isn't correct if `axis != 0` unfortunately

kasparthommen replied:
Happy to fix if you can tell me what's wrong, because all tests are green, including the ones where `axis != 0`. (There is an Azure pipeline check that fails, but that seems to be unrelated to my change as it's an [HTTP 500 internal server error](https://dev.azure.com/numpy/27346c6a-2774-4eac-bf85-e068127c0ccc/_apis/build/builds/10403/logs/44))

eric-wieser replied:
Sure, happy to elaborate, was on mobile so I thought I'd leave a breadcrumb in case other reviewers got back to it.

I think the following case fails with this patch:
```
np.geomspace(np.array([1, 2]), np.array([100, 200]), 3, axis=1)
```

kasparthommen replied:
That results in

```
[[  1.  10. 100.]
 [  2.  20. 200.]]
```
independently of whether my patch is applied or not. Note that the `moveaxis()` call is still there - *after* my code change.

eric-wieser replied:
Apologies, you're correct - this function doesn't work how I thought it did.

kasparthommen committed:
add further tests

eric-wieser commented on numpy/core/tests/test_function_base.py:
```diff
@@ -185,18 +198,34 @@ def test_start_stop_array_scalar(self):
         assert_equal(t5, t2.T)
+        for i in range(len(start)):
+            assert_(t5[i, 0] == start[i])
+            assert_(t5[i, -1] == stop[i])
```
These should all either be
```suggestion
            assert_equal(t5[i, -1], stop[i])
```
or
```suggestion
            assert t5[i, -1] == stop[i]
```
The former goes through the old numpy machinery, the new goes through the pytest machinery. Using neither results in no helpful information to show what the actual values were.

eric-wieser replied:
@charris, @mattip, any opinions on which we should be using for new tests?

kasparthommen replied:
That makes sense thanks. All newly added asserts have been changed to `assert_equal` now.

kasparthommen committed:
use assert_equal instead of assert_ to check for equality

eric-wieser commented on numpy/core/tests/test_function_base.py:
```diff
@@ -185,18 +198,34 @@ def test_start_stop_array_scalar(self):
         assert_equal(t5, t2.T)
+        for i in range(len(start)):
+            assert_(t5[i, 0], start[i])
+            assert_(t5[i, -1], stop[i])
```
```suggestion
            assert_equal(t5[i, -1], stop[i])
```
and above

kasparthommen replied:
done

eric-wieser commented on numpy/core/tests/test_function_base.py:
```diff
@@ -185,18 +198,34 @@ def test_start_stop_array_scalar(self):
         t2 = stack([geomspace(_start, _stop, 5)
                     for _start, _stop in zip(start, stop)], axis=1)
         assert_equal(t1, t2)
+        for i in range(len(start)):
+            assert_equal(t1[0, i], start[i])
+            assert_equal(t1[-1, i], stop[i])
+            assert_equal(t2[0, i], start[i])
+            assert_equal(t2[-1, i], stop[i])
```
```suggestion
```
This test isn't interesting, it's purpose is just to check that the array version behaves like the non-array.

eric-wieser replied:
Same for all of the changes in this function

kasparthommen replied:
Correct - will revert.

eric-wieser commented on numpy/core/tests/test_function_base.py:
```diff
@@ -113,6 +113,19 @@ def test_basic(self):
         assert_array_equal(y, [-100, -10, -1])
         assert_array_equal(y.imag, 0)
 
+        y = geomspace(0.3, 20.3, num=1)
```
Would be nice to explain what is being tested here.
```suggestion
    def test_exact_limits(self):
        y = geomspace(0.3, 20.3, num=1)
```

Could consider testing a negative limit too.

kasparthommen replied:
good idea, will do

kasparthommen committed:
test updates

- move test cases exposing the bug to a dedicated new test method
- undo unnecessary changes in  test_start_stop_array

eric-wieser commented on numpy/core/tests/test_function_base.py:
```diff
@@ -113,6 +113,26 @@ def test_basic(self):
 
+    def test_boundaries_match_start_and_stop_exactly(self):
+        # make sure that the boundaries of the returned array exactly
+        # equal 'start' and 'stop' - this isnt' obvious because
```
```suggestion
        # equal 'start' and 'stop' - this isn't obvious because
```

kasparthommen committed:
fix typo

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on numpy/core/tests/test_function_base.py:
```diff
@@ -113,6 +113,26 @@ def test_basic(self):
         assert_array_equal(y, [-100, -10, -1])
         assert_array_equal(y.imag, 0)
 
+    def test_boundaries_match_start_and_stop_exactly(self):
```
Could do
```suggestion
    @pytest.mark.parametrize('start_sign', [1, -1])
    @pytest.mark.parametrize('stop_sign', [1, -1])
    @pytest.mark.parametrize('num', [1, 3])
    def test_boundaries_match_start_and_stop_exactly(self, start_sign, stop_sign, num):
```
and adjust the test below accordingly

kasparthommen replied:
Note sure that improve readability...?

(Let's hope quay.io comes back online anytime soon.)

eric-wieser replied:
Ah, this actually catches an additional bug - `np.geomspace(-3, 3, num=3)` should give `[-3, nan, 3]`, not `[nan, nan, 3]`

kasparthommen replied:
True, good catch - shall I amend the unit tests accordingly?

eric-wieser replied:
Yes please!

kasparthommen replied:
Done, please review.

kasparthommen replied:
That additional test leads to a `RuntimeWarning` because it computes `log10(0.0)` which in turn seems to make the test fail. Why does a Warning lead to a test failure please? How can I avoid this?

anirudh2290 replied:
you can use something like `with np.errstate(divide="raise"):` and check that the floating point error is raised. if you want to ignore you can do something like `with np.errstate(divide="ignore"):`

kasparthommen replied:
done, thanks

kasparthommen committed:
test updates

- add test_nan_midpoint() to catch another bug that got fixed with the endpoint correction change

kasparthommen committed:
avoid RuntimeWarning

eric-wieser commented on numpy/core/tests/test_function_base.py:
```diff
@@ -134,10 +134,11 @@ def test_boundaries_match_start_and_stop_exactly(self):
-        assert_equal(y[0], -3.0)
-        assert_(isnan(y[1]))
-        assert_equal(y[2], 3.0)
+        with errstate(all='ignore'):
```
I think you can be more precise here:
```suggestion
        with errstate(invalid='ignore'):
```

kasparthommen replied:
done, thanks

eric-wieser commented on numpy/core/tests/test_function_base.py:
```diff
@@ -113,6 +113,33 @@ def test_basic(self):
+
+    def test_nan_midpoint(self):
+        with errstate(all='ignore'):
+            y = geomspace(-3, 3, num=3)
+            assert_equal(y[0], -3.0)
+            assert_(isnan(y[1]))
```
Perhaps a clearer test as:
```suggestion
            y = geomspace(-3, 3, num=4)
            assert_equal(y[0], -3.0)
            assert_(isnan(y[1:-1]).all())
```
It's not the midpoint that's nan, it's everything but the endpoint

kasparthommen replied:
also good, done

eric-wieser commented on numpy/core/tests/test_function_base.py:
```diff
@@ -113,6 +113,33 @@ def test_basic(self):
+        y = geomspace(start, stop, num=3, endpoint=False)
+        assert_equal(y[0], start)
+
+    def test_nan_midpoint(self):
```
```suggestion
    def test_nan_interior(self):
```

kasparthommen committed:
make warning catching more precise

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

kasparthommen committed:
make test more explicit

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

kasparthommen committed:
rename test to reflect the purpose more clearly

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

kasparthommen committed:
fix array index

eric-wieser commented on numpy/core/tests/test_function_base.py:
```diff
@@ -113,6 +113,33 @@ def test_basic(self):
+
+    def test_nan_interior(self):
+        with errstate(invalid='ignore'):
+            y = geomspace(-3, 3, num=4)
```
A test without the endpoint would be good too

kasparthommen replied:
good idea - done

kasparthommen committed:
add test with endpoint=False

kasparthommen commented:
@eric-wieser Any idea what I can/should do about the azure pipeline failures? Looks like they failed to download mingw builds (that do in fact exist at the specified URLs, I checked).

Can you try to re-run the failed jobs please? (I can't)

mattip commented:
That CI fail is frustrating. At first I thought it was an actual missing version, so I updated it. But now I see it is that one of the three 32-bit builds fails. I am tempted to simply remove the 3.6, 3.7 32 bit runs since they are a bit redundant. In any case I restarted the failed run.

kasparthommen commented:
@mattip Thanks, all green now, ready to merge :-)

eric-wieser commented on numpy/core/tests/test_function_base.py:
```diff
@@ -113,6 +113,37 @@ def test_basic(self):
+        assert_equal(y[0], start)
+
+    def test_nan_interior(self):
+        with errstate(invalid='ignore'):
+            y = geomspace(-3, 3, num=4)
+            assert_equal(y[0], -3.0)
+            assert_(isnan(y[1:-1]).all())
+            assert_equal(y[3], 3.0)
+
+            y = geomspace(-3, 3, num=4, endpoint=False)
+            assert_equal(y[0], -3.0)
+            assert_(isnan(y[1:]).all())
```
Could be more precise here:
```suggestion
        with errstate(invalid='ignore'):
            y = geomspace(-3, 3, num=4)
        assert_equal(y[0], -3.0)
        assert_(isnan(y[1:-1]).all())
        assert_equal(y[3], 3.0)

        with errstate(invalid='ignore'):
            y = geomspace(-3, 3, num=4, endpoint=False)
        assert_equal(y[0], -3.0)
        assert_(isnan(y[1:]).all())
```

kasparthommen replied:
done

kasparthommen committed:
more precise test

kasparthommen commented:
Just out of interest: What is the process of getting this merged into master? Will one of the reviewers do it or someone else? And using what timeline?

eric-wieser commented:
This will almost certainly make it into the 1.20 release :)

eric-wieser commented on numpy/core/tests/test_function_base.py:
```diff
@@ -113,6 +113,38 @@ def test_basic(self):
+    def test_nan_interior(self):
+        with errstate(invalid='ignore'):
+            y = geomspace(-3, 3, num=4)
+            assert_equal(y[0], -3.0)
+            assert_(isnan(y[1:-1]).all())
+            assert_equal(y[3], 3.0)
```
No warning to catch here, I think:
```suggestion
        assert_equal(y[0], -3.0)
        assert_(isnan(y[1:-1]).all())
        assert_equal(y[3], 3.0)
```

kasparthommen replied:
I'll add an empty line before the indent return as well if ok?

eric-wieser commented on numpy/core/tests/test_function_base.py:
```diff
@@ -113,6 +113,38 @@ def test_basic(self):
+
+        with errstate(invalid='ignore'):
+            y = geomspace(-3, 3, num=4, endpoint=False)
+            assert_equal(y[0], -3.0)
+            assert_(isnan(y[1:]).all())
```
or here
```suggestion
        assert_equal(y[0], -3.0)
        assert_(isnan(y[1:]).all())
```

kasparthommen replied:
dito

kasparthommen commented:
Cool, looking forward! :-)

eric-wieser reviewed (approved):
Two nits left, generally looks good to me. Feel free to just use the github UI to take the changes assuming you agree with them.

kasparthommen committed:
only keep code in the 'with' clause that actually needs it

anirudh2290 reviewed (commented):
@eric-wieser Do we need a release note for this ?
